packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx

```diff
@@ -12,19 +12,31 @@ import {Writable} from 'stream'
 
 export type WritableStream = (process: OutputProcess, index: number) => Writable
 
+export interface Footer {
+  shortcuts: Shortcut[]
+  subTitle?: string
+}
+
+export interface FooterContext {
+  footer?: Footer
+  updateShortcut: (prevShortcut: Shortcut, newShortcut: Shortcut) => void
+  updateSubTitle: (subTitle: string) => void
+}
+
 interface Shortcut {
   key: string
   action: string
+  syncer?: (footerContext: FooterContext) => void
+  metadata?: {
+    [key: string]: unknown
+  }
 }
 export interface ConcurrentOutputProps {
   processes: OutputProcess[]
   abortSignal: AbortSignal
   showTimestamps?: boolean
-  onInput?: (input: string, key: Key, exit: () => void) => void
-  footer?: {
-    shortcuts: Shortcut[]
-    subTitle?: string
-  }
+  onInput?: (input: string, key: Key, exit: () => void, footerContext: FooterContext) => Promise<void>
```

Comment:
This is breaking change so we should make sure that none is using `onInput` at Shopify or they will get compilation errors.

Reply:
Great call! We have made footerContext optional, that should make onInput backwards compatible.

Reply:
After reconsideration, we made onInputAsync as a separate optional handler in case folks want to use async methods to handle their shortcuts. This should not be a breaking change!